### PR TITLE
support kusion preview --ouput json

### LINKED ISSUE #265: [Feature Request] Kusion provides API for integration tools
## Feature Request

**Is your feature request related to a problem? Please describe:**

Now the extension is accomplishing its goals by executing the kusion binary for each time and then resolve the `stdout & stderr`, but since the stdout/stderr of kusion binary are not designed for tools integration, the output messages like this are disturbing:
image

the vscode-kusion extension requires kusion to provide APIs for tools to integrate kusion's capacities more elegantly and normatively. 

**Describe the feature you'd like:**

At least a few Go APIs of kusion are required:
- compile
- preview
- apply
- watch
- list projects
- list stacks
- list templates

### ListTemplates API

- API id：`init.ListTemplates`
- params:

| name | type | required | default value | description | example |
| ----- | ----- | -------- | ------------- | ----------- | --------- |
| online | boolean | optional | false | determines if kusion downloads templates from online repo | `true`, `false` |
| url | string | optional | `"https://github.com/KusionStack/kusion-templates"` | the online repo url, meaningful only when `online==true`|  |
| fsPath | string | optional |  | the fsPath to find the templates, meaningful only when `online == false` |`/path/to/templates` |

- returns:

| name | type | description | example |
| ----- | ----- | ----------- | -------- |
| templateList | Template[] | list of templates found | |
| err | Error | error code and message | `{"code":"invalidPath"， "message":"path not exists (or repo url not exists)"}` |

```go
type Template struct {
    name string
    description string
    content string // now the template could be the content of `kusion.yaml` file under the template directory: for example https://github.com/KusionStack/kusion-templates/blob/main/code-city/kusion.yaml
}

type Error struct {
    code  string
    message string
}
```

### PR DESCRIPTION
**Pull request by healthjyk**
#### What type of PR is this?
feature

#### What this PR does / why we need it:
when using cmd kusion preview, there is no way to get the pure preview result, cause there is prompt. Support --output json to print the preview result without any unnecessary prompt when succeeded.
#### Which issue(s) this PR fixes:

Fixes #265 

#### Does this PR introduce a user-facing change?

```release-note
the output is json marshal of struct pkg/engine/operation/models.Changes
```

#### Additional documentation e.g., design docs, usage docs, etc.:

```docs
None
```

### CODE REVIEW

File: pkg/cmd/preview/options.go [hunk 1]
```diff
@@ -72,14 +87,20 @@ func (o *PreviewOptions) Run() error {
...
 	}, project, stack)
 	if err != nil {
 		return err
 	}
 
 	// return immediately if no resource found in stack
 	if sp == nil || len(sp.Resources) == 0 {
-		fmt.Println(pretty.GreenBold("\nNo resource found in this stack."))
+		if o.Output == jsonOutput {
+			emptyChanges, _ := json.Marshal(&opsmodels.Changes{})
```
Comment (SparkYuan) [inline pkg/cmd/preview/options.go:99]: We can maintain the original output message to provide users with a friendly hint

**Review by amyXia1994 — APPROVED** [review]
lgtm

File: pkg/cmd/preview/options.go (hunk 1 above)
Comment (SparkYuan) [reply pkg/cmd/preview/options.go:99]: Maybe you can refer to the approach of Kubectl for non-existent resources. I think their approach is more user-friendly.

image

Comment (healthjyk) [reply pkg/cmd/preview/options.go]: When errored, test-format output is shown; but if succeeded, the specified format is get output.

Comment (healthjyk) [reply pkg/cmd/preview/options.go]: The left feature upgrade is that even if there is an empty spec, diff should still be done, cause it means delete.

**Comment by healthjyk** [comment]
> lgtm

@amyXia1994 if there is no resource in the planned spec, there will be no output under --output json, you may need a judgement before json unmarshal.

**Comment by healthjyk** [comment]
> lgtm



> lgtm

@amyXia1994 If there is no resource in the planned spec, there will be an empty output when --output json, you may need a judgement before json unmarshal. For the case without --output json, there is a hint, but still with successful preview.

**Review by SparkYuan — APPROVED** [review]
lgtm

**Pull request merged**